### PR TITLE
[extractor/crunchyroll] Extract other language versions (Based on Grub4K's PR)

### PR DESCRIPTION
**Pull request by ProDev2**
**IMPORTANT / MOD INFO: Only merge this PR after [Grub4K's](https://github.com/yt-dlp/yt-dlp/pull/6237) was merged.** His changes/commits are already included in this PR (because it is based on his branch). But **I AM NOT THE AUTHOR** of them and this PR must therefore wait for [his](https://github.com/Grub4K) PR to be accepted/merged. I will make sure to merge changes from [his Branch](https://github.com/Grub4K/yt-dlp/tree/fix/crunchy-various) into [mine](https://github.com/ProDev2/yt-dlp/tree/extractor/crunchyroll-new/include-formats-from-other-language-versions).

<br/>

### Description of your *pull request* and other information

Some shows on crunchyroll now have multiple audio languages attached to them e.g. [This one](https://www.crunchyroll.com/de/watch/G8WUN838K/time-will-tell). When switching to a different language in the video player, the URL changes to a different video ID. Currently the extractor only extracts the formats of one language.
My MR changes this and allows to specify languages to extract (or `all`). By default it will still only extract a single language, since extracting other ones is expensive. But when required, this can be enabled using the `language` extractor argument.
Since these languages are distributed over multiple pages you will get different `titles`, `season_ids` and `season` names. I therefore added differences in these fields to each returned format (but only if more than one language was requested/found). This way when the user requests a format (with `-f ...`) the default info dict returned will be overwritten with the correct arguments (titles, etc.) from the selected format.

**Summary: The way the extraction process works is the same as in [Grub4K's Version](https://github.com/yt-dlp/yt-dlp/pull/6237). The only thing that has really changed is that the extractor can now iterate through multiple versions and extract their formats as well if desired.**

Waiting/based on [Grub4K's PR](https://github.com/yt-dlp/yt-dlp/pull/6237)

<br/>

Details to my previous PR: I already opened a [similar PR](https://github.com/yt-dlp/yt-dlp/pull/6960) before, but that one is based on the old extractor version (prior to [Grub4K's Version](https://github.com/yt-dlp/yt-dlp/pull/6237)). I'll therefore be closing [that old PR](https://github.com/yt-dlp/yt-dlp/pull/6960) as it won't be possible to merge these older changes into the new extractor of [Grub4K](https://github.com/Grub4K) after it goes live.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code (only the commits that I made, not the ones of [Grub4K](https://github.com/Grub4K)) and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

**Comment by Jules-A** [comment]
When specifying a single language to extract it will still add all the other languages to playlist and try to extract them (in my case they got blocked from downloading by archive entry). whereas when I was previously using a filter it prevented it from trying to download.

**Comment by ProDev2** [comment]
@Jules-A Thank you. I haven't tested the playlist extractor that much yet. I'll look into it.

**Comment by Jules-A** [comment]
> @Jules-A Thank you. I haven't tested the playlist extractor that much yet. I'll look into it. Also what do you mean by 'archive entry'?

I was using download archive and they were being considered as the same entry when it tried to download them (which it shouldn't have been doing anyway). Thing is, the other languages should actually have a different id afaik so not sure why that was happening anyway.

**Comment by ProDev2** [comment]
@Jules-A The playlist extractor actually just uses the episode extractor to extract the formats. It is only there to extract the 'raw' playlist model. The playlist extractor also extracts episodes from all languages (as you may know). Therefore when setting the `language` argument you actually 'override' the language encoded in the URL given to the episode extractor (from the playlist extractor). This makes sense in code but is really confusing to the user.
I therefore changed it so that you can only set a single language in `crunchyrollbeta:language` when extracting a playlist (because this prevents weird edgecases from occurring). But it will therefore also only extract playlist items with that language.
And I also added the `crunchyrollbetashow:language` extractor argument which allows you to talk directly to the playlist extractor and request multiple languages (use this if you need to extract more that one playlist language).
If something is specified incorrectly, an error will be thrown with a message telling you what to do instead.

**TL;DR: When specifying a single language then `crunchyrollbeta:language` works for both episodes and playlists, but if you want to extract multiple languages then use `crunchyrollbetashow:language` instead.**

Hope this helps 😊

**Comment by Jules-A** [comment]
I don't know about separating the args, feels a bit weird and maybe confusing to users. I'm not even sure why it's needed? Might want to talk to Grub4k or pukkudan about it.

I haven't had time to test it but what happens if you are extracting single shows and playlists at once?

**Comment by ProDev2** [comment]
@Jules-A Every extractor has their own extractor args. I can't change the episode extractor's args from the playlist extractor (I can only read them). There is also no way to know that a playlist is being extracted from the point of view of the episode extractor. It always assumes that only one episode is being fetched. If I want to prevent the issue you mentioned, I have to prevent the episode extractor from fetching anything other than the 'default' language. Since the episode extractor args can't be changed from the playlist extractor I have to interrupt the entire extraction process if these args are set 'incorrectly'. I really see no other way of doing this effectively. This wan\'t an issue before my PR because the episode extractor always extracted the language given to it by the playlist extractor.

The only way to do things differently would be to only extract a single language in the playlist extractor when multiple languages are requested. This way the episode extractor would fetch the other language formats. But which language should be used (random?) and this would also need to take into account if `extract_flat` is set. This seems more confusing to me.

It is really rare that you want to extract a single episode and a playlist at the same time, but if you want to just set `crunchyrollbeta:language` to only a single `language code` or `default` and it will work for both episodes and playlists.

**Comment by Grub4K** [comment]
> But which language should be used (random?)

Currently we are already have the problem of metadata language, where we are guessing using account settings by doing a request to the main webpage (happens when `_update_query` gets called without a lang provided). This could maybe be stretched further to the default language here?

> Every extractor has their own extractor args. I can't change the episode extractor's args from the playlist extractor (I can only read them).

I think splitting into two extractor args is a mistake. Same as the other extractor arg it should live on `crunchyrollbeta` and affect both regular and playlist extractor. That way there will be no mismatch between the arguments, causing the afore mentioned problems. If I understand correctly it would make little sense for playlist to extract a completely different language anyways. What case am I missing here?

**Comment by ProDev2** [comment]
@Grub4K The problem rn is that for example you set the language on `crunchyrollbeta` to japanese and extract a playlist, then the playlist extractor would extract every episode from every language and would give those to the episode extractor. The episode extractor would receive the same episode but from the german, french and japanese dub and would then extract the japanese version (because it was set to it using `crunchyrollbeta:language`) three times (for each of these versions). In the info result you would get the same episode three times with the same language extracted. This happens because the episode extractor doesn't / can't know that it got the extracted urls from the playlist extractor, so it always extracts what was set using `crunchyrollbeta:language`. I think this could be really confusing to the end user. Hence why I stop the extraction process if these arguments are set 'incorrectly'. If you have any better idea on the topic I would be really glad to hear it. I also dislike the idea of having `crunchyrollbetashow` arguments. The only other way I could think of is that when you set any language other than `default` in `crunchyrollbeta:language` that only a single language is extracted in the playlist extractor. This way the episode extractor would never fetch the same format twice. But this also brings some issues with it. For example which language should be the one picked in the playlist extractor. (Sorry for bad english btw.)

**Comment by ProDev2** [comment]
Well actually, forget what I said. I have something I wanna try out tomorrow. It won't be the perfect solution (will extract some stuff twice if you set it to multiple languages or `all`), but it will be simple and will only be using `crunchyrollbeta:language`.

**Comment by Grub4K** [comment]
That makes more sense. I understand the problem now. You can use `smuggle_url` and `unsmuggle_url` from`utils` to pass data through a url

**Comment by ProDev2** [comment]
@Grub4K  Ok that sounds great. Do you think that it would make sense to extract all the requested languages for each episode in the `episode extractor` (as it is right now), or should I change it so that it only extracts a single language (the one given to it through the use of `smuggle_url`)?

**Comment by Grub4K** [comment]
After playlist extraction, if we always just let the episode extractor extract a single language, what edge cases could exist? I cant think of any. If thats the case, then that would be the best way to handle this:
- Playlist extractor extracts individual urls, one for each language
- Creates a url_result with smuggled data to only extract single language
- Episode extractor unsmuggles and extracts single language only

Let me know what you think

**Comment by ProDev2** [comment]
@Grub4K I think its way better now. Using `smuggle_url` really helped. 👍 Thank you!

**Comment by Jules-A** [comment]
> @Jules-A Is it more to your liking now? I changed it so that you only need to specify `crunchyrollbeta:language` for both playlist and episode extraction. I also think it's easier to use this way.

I haven't had time to test it properly look through the code but yeah, it's much better like that.

As for how I would prefer things be done:
If I use `--extractor-args 'language=en-US,ja-JP'` I'd want it to try to extract en-US but if it doesn't exist, try ja-JP. However the comma seems to suggest it could try downloading both. Honestly I don't mind if it tries downloading both since I could just use `-f '(bv*+ba/b)[language=en-US]/(bv*+ba/b)'` to just get the format I want but it might not be clear to most users. It would honestly be pretty cool if we could do something like `--extractor-args 'language=en-US|ja-JP'` but I don't know how pukkudan would feel about that.

**Comment by Grub4K** [comment]
While I certainly did not think of A or B as a usecase of this, we can get something similar, however not as performant, by specifying `language=en-US,ja-JP` and then **also** using `(bv*+ba/b)[language=en-US]/(bv*+ba/b)`. While that will extract jp unnecessarily, it will at least give the desired result...

**Comment by ProDev2** [comment]
@Grub4K Is there a way to `unsmuggle` the url before returning it to the user? Because right now when extracting a playlist (and `flat_playlist` is set) you will get the smuggled data as a result in the `url` field. This could potentially break other user\'s code. Is there a way ro do some sort of 'post processing' on the `info dict` or `url` before it is returned?

**Comment by pukkandan** [comment]
Smuggled data is in fragment part of URL and therefore should not break anything.

**Comment by ProDev2** [comment]
@pukkandan Ok thank you! Good to know. 👍

**Comment by pukkandan** [comment]
I agree with https://github.com/yt-dlp/yt-dlp/pull/7009#issuecomment-1554197289

**Comment by Rafawell** [comment]
--match-filters "language=pt-BR" works very well for playlists, although it doesn't work with the URL of a single episode.

**Comment by Jules-A** [comment]
Seems like way too much for a specific extractor option. If all extractors had options like this then the readme page would be an absolute mess. If you are going to put this much work in to it then maybe it's worth asking pukkudan if it could be used as a global option that other extractors could possibly also use? Then if that's considered I guess you could also get the language selection from format selection but pretty sure that would be extremely hard to do.

**Comment by ProDev2** [comment]
@Jules-A Thank you for your kind words. It would be cool if it could be used someway. But even if not, I did learn a lot. So it's ok. Had a lot of fun making it.

**Comment by ProDev2** [comment]
Don\'t know why, but I wasn\'t able to merge the master into my branch, even though it was based on the latest commit of Grub4K\'s PR. But I merged it to resolve the conflicts.

**Comment by Grub4K** [comment]
This was most likely bc my PRs are squashed on merge. Since this PR will also be squashed it will be fine and you shouldnt have to worry. I will get to properly reviewing this PR soon

**Comment by ProDev2** [comment]
Crunchyroll extractor is currently broken because `_download_webpage` returns a `HTTP Error 403: Forbidden` error when trying to  download `https://www.crunchyroll.com/de/` (At least for me). Really strange. Does anyone know why?

**Comment by Jules-A** [comment]
Done some very limited testing on free content (don't have sub atm) and seems to work fine. What's the hold up on this PR?

**Comment by Grub4K** [comment]
A lot on my table personally and for this project. I feel like this PR is overly complex and disecting how it works and could be improved will be a lot of work. I will get to it though.

**Comment by Grub4K** [comment]
Note that we can't merge formats per language variant, as per my comment ([#7741 (comment)](<https://github.com/yt-dlp/yt-dlp/issues/7741#issuecomment-1732669953>)). If that is the goal it would make a great plugin instead. I'm open to help simplifying your work by having the extractor expose more functions for the plugin to override.

If you want to filter by language you can already use `--match-filter language=...`. The extractor arg makes sense only for reducing network calls on season extraction. That could be implemented like so: [Grub4K/yt-dlp@a3f7be352](<https://github.com/Grub4K/yt-dlp/commit/a3f7be352dbe02bcf8d19e566f7eaa63faa82dce>)

**Comment by ProDev2** [comment]
@Grub4K Sorry didn't see your message. Wouldn't [Grub4K/yt-dlp@96b2919](https://github.com/Grub4K/yt-dlp/commit/96b2919146b9b7b06c486d979cf7fe03ce8d7b64) only work for playlists though? And what exactly is a plugin (-> how do you make or use one?) I have no problem rewriting my code as a plugin (and maybe making it shorter in the process as well). Also is installig a plugin complicated?

**Comment by dishan009** [comment]
Hello prodev i use this command yt-dlp --audio-quality 0 --audio-multistreams --embed-subs  --write-subs  --sub-langs all --remux-video mkv --extractor-args crunchyrollbeta:language=en-US,ja-JP --username EmperorFirmPaladin https://www.crunchyroll.com/watch/GX9UQ5DE7/deep-sleep

But it’s only downloading the English dub. Please help me with the problem.

**Comment by Jules-A** [comment]
>www.crunchyroll.com/watch/GX9UQ5DE7/deep-sleep

Have you even checked with -F that they are available? That page just gives me a 404 so I have no clue even what show you are trying to dl... Might need to say what region.

**Comment by dishan009** [comment]
Hello, this command → yt-dlp -F --extractor-args crunchyrollbeta:language=all --cookies cookies.txt https://www.crunchyroll.com/watch/G7PU4MZ1G/the-journeys-end    is showing me this →  ID                                EXT RESOLUTION FPS │   FILESIZE   TBR PROTO │ VCODEC      ACODEC    MORE INFO
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
adaptive_hls-hardsub-ar-SA-meta   mp4 multiple       │                  m3u8F │ unknown     unknown   [ja-JP] Quality selection URL
adaptive_hls-hardsub-de-DE-meta   mp4 multiple       │                  m3u8F │ unknown     unknown   [ja-JP] Quality selection URL
adaptive_hls-hardsub-en-US-meta-0 mp4 multiple       │                  m3u8F │ unknown     unknown   [en-US] Quality selection URL
adaptive_hls-hardsub-en-US-meta-1 mp4 multiple       │                  m3u8F │ unknown     unknown   [ja-JP] Quality selection URL
adaptive_hls-hardsub-es-419-meta  mp4 multiple       │                  m3u8F │ unknown     unknown   [ja-JP] Quality selection URL
adaptive_hls-hardsub-es-ES-meta   mp4 multiple       │                  m3u8F │ unknown     unknown   [ja-JP] Quality selection URL
adaptive_hls-hardsub-fr-FR-meta   mp4 multiple       │                  m3u8F │ unknown     unknown   [ja-JP] Quality selection URL
adaptive_hls-hardsub-hi-IN-meta   mp4 multiple       │                  m3u8F │ unknown     unknown   [ja-JP] Quality selection URL
adaptive_hls-hardsub-it-IT-meta   mp4 multiple       │                  m3u8F │ unknown     unknown   [ja-JP] Quality selection URL
adaptive_hls-hardsub-pt-BR-meta   mp4 multiple       │                  m3u8F │ unknown     unknown   [ja-JP] Quality selection URL
adaptive_hls-hardsub-ru-RU-meta   mp4 multiple       │                  m3u8F │ unknown     unknown   [ja-JP] Quality selection URL
adaptive_hls-562-0                mp4 428x240     24 │ ~ 94.43MiB  563k m3u8  │ avc1.42c015 mp4a.40.2 [ja-JP]
adaptive_hls-562-1                mp4 428x240     24 │ ~ 94.43MiB  563k m3u8  │ avc1.42c015 mp4a.40.2 [en-US]
adaptive_hls-1093-0               mp4 640x360     24 │ ~183.49MiB 1093k m3u8  │ avc1.4d401e mp4a.40.2 [ja-JP]
adaptive_hls-1093-1               mp4 640x360     24 │ ~183.52MiB 1093k m3u8  │ avc1.4d401e mp4a.40.2 [en-US]
adaptive_hls-2092-0               mp4 848x480     24 │ ~351.22MiB 2092k m3u8  │ avc1.4d401f mp4a.40.2 [ja-JP]
adaptive_hls-2092-1               mp4 848x480     24 │ ~351.23MiB 2092k m3u8  │ avc1.4d401f mp4a.40.2 [en-US]
adaptive_hls-4116                 mp4 1280x720    24 │ ~690.96MiB 4116k m3u8  │ avc1.640028 mp4a.40.2 [ja-JP]
adaptive_hls-4118                 mp4 1280x720    24 │ ~691.32MiB 4118k m3u8  │ avc1.640028 mp4a.40.2 [en-US]
adaptive_hls-8091                 mp4 1920x1080   24 │ ~  1.33GiB 8091k m3u8  │ avc1.640028 mp4a.40.2 [ja-JP]
adaptive_hls-8108                 mp4 1920x1080   24 │ ~  1.33GiB 8108k m3u8  │ avc1.640028 mp4a.40.2 [en-US]


So both english and japanese dub is available but only english dub is downloading and nothing else and by the way i am using a premium account. I want jp and en audio merge in a single video.

**Comment by dishan009** [comment]
And this command → yt-dlp --extractor-args crunchyrollbeta:language=all --cookies cookies.txt https://www.crunchyroll.com/watch/GX9UQ5DE7/deep-sleep full result is→ [crunchyroll] Extracting URL: https://www.crunchyroll.com/watch/GX9UQ5DE7/deep-sleep
[crunchyroll] Authenticating with grant_type=etp_rt_cookie
[crunchyroll] GX9UQ5DE7: Downloading object info JSON (cms)
[crunchyroll] G8WUN2J15: Downloading object info JSON (cms)
[crunchyroll] G8WUNWNG2: Downloading object info JSON (cms)
[crunchyroll] GG1U2EE33: Downloading object info JSON (cms)
[crunchyroll] GWDU8G5PJ: Downloading object info JSON (cms)
[crunchyroll] GPWUK4228: Downloading object info JSON (cms)
[crunchyroll] G8WUN8QDQ: Downloading object info JSON (cms)
[crunchyroll] G4VUQN35Z: Downloading object info JSON (cms)
[crunchyroll] G9DUE85Q4: Downloading object info JSON (cms)
[crunchyroll] G8WUNZG8M: Downloading object info JSON (cms)
[crunchyroll] G9DUE5E0Q: Downloading object info JSON (cms)
[crunchyroll] G4VUQ14QJ: Downloading object info JSON (cms)
[crunchyroll] GQJUGV1X8: Downloading object info JSON (cms)
[crunchyroll] Retrieving signed policy
[crunchyroll] GX9UQ5DE7: Downloading stream info JSON (signed cms)
[crunchyroll] To get all formats of a hardsub language, use "--extractor-args crunchyrollbeta:hardsub=<language_code or all>". See https://github.com/yt-dlp/yt-dlp#crunchyrollbeta-crunchyroll for more info
[crunchyroll] GX9UQ5DE7: Downloading adaptive_hls HLS manifest
[crunchyroll] GX9UQ5DE7: Downloading chapter info
[crunchyroll] G8WUN2J15: Downloading stream info JSON (signed cms)
[crunchyroll] G8WUN2J15: Downloading adaptive_hls HLS manifest
[crunchyroll] G8WUN2J15: Downloading chapter info
[crunchyroll] G8WUNWNG2: Downloading stream info JSON (signed cms)
[crunchyroll] G8WUNWNG2: Downloading adaptive_hls HLS manifest
[crunchyroll] G8WUNWNG2: Downloading chapter info
[crunchyroll] GG1U2EE33: Downloading stream info JSON (signed cms)
[crunchyroll] GG1U2EE33: Downloading adaptive_hls HLS manifest
[crunchyroll] GG1U2EE33: Downloading chapter info
[crunchyroll] GWDU8G5PJ: Downloading stream info JSON (signed cms)
[crunchyroll] GWDU8G5PJ: Downloading adaptive_hls HLS manifest
[crunchyroll] GWDU8G5PJ: Downloading chapter info
[crunchyroll] GPWUK4228: Downloading stream info JSON (signed cms)
[crunchyroll] GPWUK4228: Downloading adaptive_hls HLS manifest
[crunchyroll] GPWUK4228: Downloading chapter info
[crunchyroll] G8WUN8QDQ: Downloading stream info JSON (signed cms)
[crunchyroll] G8WUN8QDQ: Downloading adaptive_hls HLS manifest
[crunchyroll] G8WUN8QDQ: Downloading chapter info
[crunchyroll] G4VUQN35Z: Downloading stream info JSON (signed cms)
[crunchyroll] G4VUQN35Z: Downloading adaptive_hls HLS manifest
[crunchyroll] G4VUQN35Z: Downloading chapter info
[crunchyroll] G9DUE85Q4: Downloading stream info JSON (signed cms)
[crunchyroll] G9DUE85Q4: Downloading adaptive_hls HLS manifest
[crunchyroll] G9DUE85Q4: Downloading chapter info
[crunchyroll] G8WUNZG8M: Downloading stream info JSON (signed cms)
[crunchyroll] G8WUNZG8M: Downloading adaptive_hls HLS manifest
[crunchyroll] G8WUNZG8M: Downloading chapter info
[crunchyroll] G9DUE5E0Q: Downloading stream info JSON (signed cms)
[crunchyroll] G9DUE5E0Q: Downloading adaptive_hls HLS manifest
[crunchyroll] G9DUE5E0Q: Downloading chapter info
[crunchyroll] G4VUQ14QJ: Downloading stream info JSON (signed cms)
[crunchyroll] G4VUQ14QJ: Downloading adaptive_hls HLS manifest
[crunchyroll] G4VUQ14QJ: Downloading chapter info
[crunchyroll] GQJUGV1X8: Downloading stream info JSON (signed cms)
[crunchyroll] GQJUGV1X8: Downloading adaptive_hls HLS manifest
[crunchyroll] GQJUGV1X8: Downloading chapter info
[info] GX9UQ5DE7: Downloading 1 format(s): adaptive_hls-8115-1
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 343
[download] Destination: Demon Slayer： Kimetsu no Yaiba Mugen Train Arc (English-IN Dub) Episode 2 – Deep Sleep [G4VUQN35Z].mp4
[download]   2.7% of ~   1.25GiB at  413.82KiB/s ETA 01:09:59 (frag 9/343)

**Comment by Grub4K** [comment]
> Wouldn't [Grub4K/yt-dlp@96b2919](https://github.com/Grub4K/yt-dlp/commit/96b2919146b9b7b06c486d979cf7fe03ce8d7b64) only work for playlists though?

Correct, that is the only reasonable way to work within the constraints of the yt-dlp extractor system. Since each audio version has its own url and id, each should also be downloaded separately.
> And what exactly is a plugin (-> how do you make or use one?)

A plugin is something like an extractor that gets dynamically loaded. Plugins dont go through our review process and are also mot required to follow strict guidelines (like the one mentioned above). You can read up on creating a plugin [here](<https://github.com/yt-dlp/yt-dlp#developing-plugins>)
> Also is installig a plugin complicated?

No, you can install it by placing a python file in specific locations. Read more about it [here](<https://github.com/yt-dlp/yt-dlp#installing-plugins>)

**Comment by Grub4K** [comment]
@dishan009 that is what this PR is wanting to fix, but as i have pointed out its not quite how yt-dlp is suppose to work, and it also cant work using audio-multistreams. Create a wrapper script instead and merge them afterwards

**Comment by ProDev2** [comment]
@Grub4K Thanks a lot. I'll look into how to turn it into a plugin soon (once I have more time than rn). For now should I close this PR? Because Playlist extraction is so fast anyway that there is almost no point in pre-filtering it.

**Comment by Grub4K** [comment]
If you agree with the reasoning I gave in my comment feel free to close this PR. Thank you for understanding!

**Comment by ProDev2** [comment]
@Grub4K I agree. I'll close this now, but I will add further comments in here once I have a working plugin.
